File: prediteq_frontend/src/lib/costModel.ts
import type { TacheType } from "@/lib/runtimeDataRepository";

export const LABOR_RATE_PER_HOUR = 35;

const TASK_COST_ASSUMPTIONS: Record<
  TacheType,
  { laborHours: number; partsCost: number; label: string }
> = {
  preventive: {
    laborHours: 4,
    partsCost: 140,
    label: "Visite preventive",
  },
  inspection: {
    laborHours: 2,
    partsCost: 70,
    label: "Inspection terrain",
  },
  corrective: {
    laborHours: 6,
    partsCost: 410,
    label: "Intervention corrective ciblee",
  },
};

export function getTaskCostReference(type: TacheType) {
  const assumption = TASK_COST_ASSUMPTIONS[type];
  const laborCost = assumption.laborHours * LABOR_RATE_PER_HOUR;
  const totalCost = laborCost + assumption.partsCost;

  return {
    ...assumption,
    laborRate: LABOR_RATE_PER_HOUR,
    laborCost,
    totalCost,
  };
}

export function getTaskBaselineCost(type: TacheType) {
  return getTaskCostReference(type).totalCost;
}

export function getBudgetReferenceCost(type: TacheType, historicalAverage = 0) {
  return Math.max(historicalAverage, getTaskBaselineCost(type));
}

File: prediteq_api/core/cost_model.py
from __future__ import annotations

from typing import Literal

TaskType = Literal["preventive", "inspection", "corrective"]

LABOR_RATE_PER_HOUR = 30

TASK_COST_ASSUMPTIONS: dict[TaskType, dict[str, int | str]] = {
    "preventive": {
        "labor_hours": 2,
        "parts_cost": 200,
        "label": "Visite preventive",
    },
    "inspection": {
        "labor_hours": 3,
        "parts_cost": 230,
        "label": "Inspection renforcee",
    },
    "corrective": {
        "labor_hours": 6,
        "parts_cost": 300,
        "label": "Intervention corrective",
    },
}


def get_task_cost_reference(task_type: TaskType) -> dict[str, int | str]:
    assumption = TASK_COST_ASSUMPTIONS[task_type]
    labor_hours = int(assumption["labor_hours"])
    parts_cost = int(assumption["parts_cost"])
    labor_cost = labor_hours * LABOR_RATE_PER_HOUR

    return {
        **assumption,
        "labor_rate": LABOR_RATE_PER_HOUR,
        "labor_cost": labor_cost,
        "total_cost": labor_cost + parts_cost,
    }


def get_task_baseline_cost(task_type: TaskType) -> int:
    return int(get_task_cost_reference(task_type)["total_cost"])


def get_budget_reference_cost(task_type: TaskType, historical_average: float = 0.0) -> float:
    return max(float(historical_average), float(get_task_baseline_cost(task_type)))
